core/src/main/java/org/apache/spark/io/ReadAheadInputStream.java

```diff
@@ -232,7 +229,9 @@ private void waitForAsyncReadComplete() throws IOException {
     stateChangeLock.lock();
     try {
       while (readInProgress) {
+        isWaiting.set(true);
         asyncReadComplete.await();
+        isWaiting.set(false);
```

Comment:
What happens if `await()` throws an exception? Is it ok not to update `isWaiting`?

Reply:
Good catch, I added `isWaiting.set(false)` to the finally branch.
Actually, since the whole implementation assumes that there is only one reader, I removed the while() loop, since there is no other reader to race with us to trigger another read.

In practice I think not updating `isWaiting` it would have been benign, as after the exception the query will be going down with an `InterruptedException`, or elsewise anyone upstream handling that exception would most probably declare that stream as unusable afterwards anyway.